src/lua/core/Import.cpp

```diff
@@ -60,390 +104,393 @@ static std::string get_caller(lua_State *L)
+	size_t end = path.find_first_of('/', 1); // use start=1 to generate string::npos when the path == "/"
+
+	while (end != std::string::npos) {
+		auto fragment = path.substr(start, end - start);
+
+		if (!fragment.empty()) {
+			// Paths containing the separator character cannot be translated into a module name
+			// this means you can't do implicit lookup using a path starting in a hidden directory, e.g.
+			// /data/.local/mylib.lua
+			if (fragment.find('.') != std::string::npos)
+				return "";
+			module_name.append(fragment).push_back('.');
 		}
 
-		if (!isImported && !importInfo.isFullName) {
-			cacheName = FileSystem::JoinPath(cacheName, "init.lua");
-			isImported = get_cached(L, cacheName);
-		}
-
-		if (isImported) {
-			DEBUG_PRINTF(" found cached %s\n", cacheName.c_str());
-			break;
-		}
+		start = end + 1;
+		end = path.find_first_of('/', start);
 	}
 
-	if (!isImported)
-		DEBUG_PRINTF(" not cached\n");
+	if (module_name.back() == '.')
+		module_name.pop_back();
 
-	LUA_DEBUG_END(L, (isImported ? 1 : 0));
-	return isImported;
+	return module_name;
 }
 
-/**
-* Tries to load file.
-* Returns true if file exists and puts return of dofile to stack.
-*/
-static bool load_file(lua_State *L, const std::string &name)
+// If the given file path is cached, returns the cached object on the stack.
+// Otherwise if the path points to a valid file, executes that file and sets
+// the result in the module cache.
+// In all cases, if this function returns false the stack is unmodified.
+static bool load_file_cached(lua_State *L, const std::string &path, int cacheIdx)
 {
 	LUA_DEBUG_START(L);
+	cacheIdx = lua_absindex(L, cacheIdx);
 
-	RefCountedPtr<FileSystem::FileData> fileData = FileSystem::gameDataFiles.ReadFile(name);
-
-	// if file exists
-	if (fileData) {
-		DEBUG_PRINTF(" loading %s...\n", name.c_str());
-		pi_lua_dofile(L, *fileData, 1);
-	}
-
-	LUA_DEBUG_END(L, (fileData ? 1 : 0));
-	return static_cast<bool>(fileData);
-}
-
-/**
-* Tries to load file.
-* If file found returns true, puts return of dofile to stack, no matter nil it or value.
-* If import was successful, changes fileName in the importInfo to the filename that was found.
-*/
-static bool import_from_file(lua_State *L, ImportInfo &importInfo)
-{
-	LUA_DEBUG_START(L);
-	DEBUG_INDENTED_PRINTF("import [%s]: trying to load a file...", importInfo.importName.c_str());
+	// If the object is in the import cache, return it.
+	lua_getfield(L, cacheIdx, path.c_str());
+	if (!lua_isnil(L, -1)) {
+		DEBUG_INDENTED_PRINTF("-> loaded module file %s from cache.\n", path.c_str());
+		LUA_DEBUG_END(L, 1);
+		return true;
+	} else
+		lua_pop(L, 1);
 
-	bool isImported = false;
-	std::string realName = importInfo.importName;
+	// FileSystem::JoinPathBelow can't join a path starting with '/' for some reason
+	// TODO: find out why and remove that restriction.
+	std::string filePath = (path[0] == '/') ? path.substr(1) : path;
+	auto fileInfo = FileSystem::gameDataFiles.Lookup(filePath);
```

Comment:
That would be because a leading `/`is interpreted as the file system's root folder, and is only allowed if the basepath is set to in `src/FileSystem.cpp` is also set to root. However we explicitly set it to the game's data folder, and user data folder on init so we don't end up crawling all over the file system.

Reply:
The semantics behind that decision boggle my mind though; the name `JoinPathBelow` reads as though the function was purposefully *intended* to translate an absolute virtual-root path into a real filesystem path, and not throw `std::invalid_argument` if passed a non-relative second path.

Oh well, just another reason to re-design the interface of the filesystem abstraction. Just pin that on the docket in about five year's time... :D